### DISCUSSION TITLE
Timers

### DISCUSSION THREAD

**Discussion by rahav** [contiki-ng/contiki-ng#1679] [Q&A] [answered]
I'm learning how to use [Contiki timers](https://github.com/contiki-ng/contiki-ng/wiki/Documentation:-Timers). The documentation is silent about when the function that is invoked when the timer elapses should be using `PROCESS_BEGIN();` and `PROCESS_END();`.

Is there any formal definition of when this is needed?

**Comment by atiselsts** [comment]
Can you rephrase the question or explain the problem you're encountering? I'm afraid I did not understand it.

**Reply by rahav** [reply]
Hi @atiselsts and thanks for the quick response. It comprehension on my part. When do I need to use `PROCESS_BEGIN();` and `PROCESS_END();`? Or what is gained when using it?

**Comment by atiselsts** [comment]
These macros are used in Contiki process protothreads. Here's the documentation related to processes: https://github.com/contiki-ng/contiki-ng/wiki/Documentation:-Processes-and-events

The connection between processes and timers is via event timers. When an events timer expires, a system event is generated. Contiki-NG process can wait for that event and execute custom code when it happens.

**Reply by rahav** [reply]
I've read it. It is still not clear when we need a protothread.

In the Timers documentation, the sections on [Timer Library](https://github.com/contiki-ng/contiki-ng/wiki/Documentation:-Timers#the-timer-library), [Ctimer Library](https://github.com/contiki-ng/contiki-ng/wiki/Documentation:-Timers#the-ctimer-library), and [Rtimer Library](https://github.com/contiki-ng/contiki-ng/wiki/Documentation:-Timers#the-rtimer-library) -- do NOT use them.

But the section on [The Etimer Library](https://github.com/contiki-ng/contiki-ng/wiki/Documentation:-Timers#the-etimer-library) does.

So, at least in my head, it is not clear when do we need to use a protothread? Or what is the benefits of using one, if the examples show that a timer can work without a protothread?

**Comment by atiselsts** [comment, marked as answer]
Etimer is short for "event timer". Event timers are special in the way that they generate process events, as explained above.

You can get the timer functionality in your application without adding any new processes to your application (process = protothread for the sake of this example). But a Contiki-NG application requires that there is at least one user-defined process. This requirement is the main reason why you should use (at least one) process. To get deeper than this you need to read up computer science concepts, such as thread-based vs. event-based approaches for the execution flow and what does waiting for an event means.
